rofl-client/ts/src/roflClient.ts

```diff
@@ -0,0 +1,284 @@
+
+        if (!res || typeof res !== 'object' || res === null) {
+            throw new Error('Invalid response from ROFL metadata');
+        }
+
+        return res as Record<string, string>;
+    }
+
+    /** Replace metadata key-value pairs (triggers registration refresh if changed). */
+    async setMetadata(metadata: Record<string, string>): Promise<void> {
+        await this.appdRequest('POST', '/rofl/v1/metadata', metadata);
+    }
+
+    /** Convenience: Returns ROFL app ID (bech32). */
+    async getAppId(): Promise<string> {
+        const res = await this.appdRequest('GET', '/rofl/v1/app/id');
+
+        if (typeof res !== 'string') {
+            throw new Error('Invalid response while fetching ROFL app id');
+        }
+
+        return res;
+    }
+
+    // internals
+
+    private isHttpUrl(): boolean {
+        return this.url.startsWith('http://') || this.url.startsWith('https://');
+    }
+
+    private socketPathOrDefault(): string {
+        if (this.url && !this.isHttpUrl()) return this.url;
+        if (!this.url) return ROFL_SOCKET_PATH;
+        return '';
+    }
+
+    private async appdRequest(
+        method: 'GET' | 'POST',
+        path: string,
+        payload?: unknown,
+    ): Promise<unknown> {
+        const headers: Record<string, string> = {
+            'User-Agent': this.userAgent,
+        };
+        let body: string | undefined;
+
+        if (method === 'POST') {
+            body = JSON.stringify(payload ?? {});
+            // Canonical header casing for broader compatibility.
+            headers['Content-Type'] = 'application/json';
+            headers['Content-Length'] = Buffer.byteLength(body).toString();
+        }
+
+        const transportRequest: TransportRequest = this.isHttpUrl()
+            ? {
+                  method,
+                  path,
+                  baseUrl: this.url,
+                  payload,
+                  body,
+                  timeoutMs: this.timeoutMs,
+                  headers,
+              }
+            : {
+                  method,
+                  path,
+                  socketPath: this.socketPathOrDefault(),
+                  payload,
+                  body,
+                  timeoutMs: this.timeoutMs,
+                  headers,
+              };
+
+        const transport = this.transport ?? nodeTransport;
+        const {status, body: responseBody} = await transport(transportRequest);
+
+        if (status >= 400) {
+            const message =
+                typeof responseBody === 'string' ? responseBody : JSON.stringify(responseBody);
+            throw new Error(`ROFL request failed (${status}): ${message}`);
+        }
+
+        return responseBody;
+    }
+}
+
+/** Default Node.js transport with UDS + HTTP(S) support. */
+const nodeTransport: Transport = (req) =>
+    new Promise<TransportResponse>((resolve, reject) => {
+        const isHttp = !!req.baseUrl;
+        const isHttps = isHttp && req.baseUrl!.startsWith('https://');
```

Comment:
Is there some way to run rofl apps that makes /run/rofl-appd.sock use TLS?

Reply:
appd only binds to the Unix socket, so in the default deployment, nothing leaves the host and TLS isn't involved. the HTTP/HTTPS part in the client (also in Python) is just for advanced deployments where operators choose to front the socket with a TCP/TLS proxy.